### PR TITLE
update CMake min version to 3.15

### PR DESCRIPTION
**Pull request by vhvb1989**
Storage is using CMake version 3.15 and it is currently required for generating without errors

### CODE REVIEW

**Review by RickWinter — COMMENTED** [review]
Although I'd love to be on 3.15, its a very recent build (less than a year).  Is there a reason for requiring 3.15?

**Comment by vhvb1989** [comment]
> 
> 
> Although I'd love to be on 3.15, its a very recent build (less than a year). Is there a reason for requiring 3.15?

Storage CMake project is already using 3.15
https://github.com/Azure/azure-sdk-for-cpp/blob/master/sdk/storage/CMakeLists.txt#L4

First I tried to change everything to 3.12 but I got this error
```cmake
-- The CXX compiler identification is GNU 9.2.1
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The C compiler identification is GNU 9.2.1
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Found PythonInterp: /usr/bin/python (found version "2.7.17") 
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found OpenSSL: /home/vivazqu/workspace/vcpkg/installed/x64-linux/debug/lib/libcrypto.a (found suitable version "1.1.1g", minimum required is "1")  
-- Found ZLIB: /home/vivazqu/workspace/vcpkg/installed/x64-linux/debug/lib/libz.a (found suitable version "1.2.11", minimum required is "1") 
-- Retrieving version from /home/vivazqu/workspace/azure-sdk-for-cpp/sdk/core/azure-core/inc/azure/core/version.hpp
-- VERSION_MAJOR 1
-- VERSION_MINOR 0
-- VERSION_PATCH 0
-- VERSION_PRERELEASE beta.2
-- AZ_LIBRARY_VERSION 
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- Found LibXml2: /home/vivazqu/workspace/vcpkg/installed/x64-linux/debug/lib/libxml2.a (found version "2.9.9") 
-- Found ZLIB: /home/vivazqu/workspace/vcpkg/installed/x64-linux/debug/lib/libz.a (found version "1.2.11") 
-- Found OpenSSL: /home/vivazqu/workspace/vcpkg/installed/x64-linux/debug/lib/libcrypto.a (found version "1.1.1g")  
Azure Storage Common Package Version 1.0.0-beta.2
Azure Storage Blobs Package Version 1.0.0-beta.2
CMake Error at sdk/storage/azure-storage-blobs/CMakeLists.txt:66 (target_link_libraries):
  Attempt to add link library "azure::storage::blobs" to target
  "azure-storage-test" which is not built in this directory.

  This is allowed only when policy CMP0079 is set to NEW.


CMake Error at sdk/storage/azure-storage-blobs/CMakeLists.txt:74 (target_link_libraries):
  Attempt to add link library "azure::storage::blobs" to target
  "azure-storage-sample" which is not built in this directory.

  This is allowed only when policy CMP0079 is set to NEW.


Azure Storage Files DataLake Package Version 1.0.0-beta.2
CMake Error at sdk/storage/azure-storage-files-datalake/CMakeLists.txt:62 (target_link_libraries):
  Attempt to add link library "azure::storage::files::datalake" to target
  "azure-storage-test" which is not built in this directory.

  This is allowed only when policy CMP0079 is set to NEW.


CMake Error at sdk/storage/azure-storage-files-datalake/CMakeLists.txt:70 (target_link_libraries):
  Attempt to add link library "azure::storage::files::datalake" to target
  "azure-storage-sample" which is not built in this directory.

  This is allowed only when policy CMP0079 is set to NEW.


Azure Storage Files Shares Package Version 1.0.0-beta.2
CMake Error at sdk/storage/azure-storage-files-shares/CMakeLists.txt:56 (target_link_libraries):
  Attempt to add link library "azure::storage::files::shares" to target
  "azure-storage-test" which is not built in this directory.

  This is allowed only when policy CMP0079 is set to NEW.


CMake Error at sdk/storage/azure-storage-files-shares/CMakeLists.txt:64 (target_link_libraries):
  Attempt to add link library "azure::storage::files::shares" to target
  "azure-storage-sample" which is not built in this directory.

  This is allowed only when policy CMP0079 is set to NEW.


-- Retrieving version from /home/vivazqu/workspace/azure-sdk-for-cpp/sdk/template/azure-template/inc/azure/template/version.hpp
-- VERSION_MAJOR 1
-- VERSION_MINOR 0
-- VERSION_PATCH 0
-- VERSION_PRERELEASE beta.7
-- AZ_LIBRARY_VERSION 
-- Configuring incomplete, errors occurred!
See also "/home/vivazqu/workspace/azure-sdk-for-cpp/build/CMakeFiles/CMakeOutput.log".
See also "/home/vivazqu/workspace/azure-sdk-for-cpp/build/CMakeFiles/CMakeError.log".
```

That's why I went to better upgrade all to 3.15

The alternative is to keep in 3.12 and find out how to fix those CMake issues from Storage

**Comment by vhvb1989** [comment]
Found this:
https://cmake.org/cmake/help/git-stage/policy/CMP0079.html

We require at least 3.13 due to this

**Comment by ahsonkhan** [comment]
Anyway to workaround this (possibly by re-writing part of the cmakelist file) and only require cmake 3.10 (the default on Ubuntu 16.04)?

**Comment by vhvb1989** [comment]
Found this:
https://cmake.org/cmake/help/git-stage/policy/CMP0079.html

We require at least 3.13 due to this

> 
> 
> Anyway to workaround this (possibly by re-writing part of the cmakelist file) and only require cmake 3.10 (the default on Ubuntu 16.04)?

Wonder if it is worth it...
We would need to refactor storage unit test CMake-project.  
Storage Cmake project is currently defining  storage-test CMake target on its root CMake list, then each subfolder adds source files to this target (blobs, files, etc).  This is only possible for CMake 3.13
Something like
```cmake
#root cmakelist
add_executable(storage-test)
# then add subdirs
add_subdir(blob)
add_subdir(fliles)
..
..

#then from /blob -> CMakeList:
target_sources(storage-test PRIVATE 
src/test
src/test2...)

#then from /file -> CMakeList:
target_sources(storage-test PRIVATE 
src/test
src/test2...)
```
We would need to refactor the root CMake list to add all src/tests directly like
```cmake
#root cmakelist
add_executable(storage-test PRIVATE
blobs/test/src/test
blobs/test/src/test2
files/test/src/test
files/test/src/test2)

# Then remove the target_sources from subfolders
```

For Azure Core, we follow the second approach, all source tests are added on the same file where the cmake target is defined.

IMO, we should not consider ubuntu 16 anymore. It will reach to End of life next year (apr 2021).  I would like to consider 18 and 20 better.
Chances of someone picking Ubuntu 16 as a fresh installation are low.  And current Ubuntu 16 users should have already update apt-get to get something newer

**Comment by ahsonkhan** [comment]
> IMO, we should not consider ubuntu 16 anymore. It will reach to End of life next year (apr 2021). I would like to consider 18 and 20 better.

That's a fair point. However...
Can you double check what is the default cmake version for 18.04 and 20.04?

I read somewhere, it is still 3.10.2, which adds a friction point:
https://graspingtech.com/upgrade-cmake/
> The version of CMake installed by APT on Ubuntu 18.04 is currently 3.10.2.

**Comment by vhvb1989** [comment]
> 
> 
> > IMO, we should not consider ubuntu 16 anymore. It will reach to End of life next year (apr 2021). I would like to consider 18 and 20 better.
> 
> That's a fair point. However...
> Can you double check what is the default cmake version for 18.04 and 20.04?
> 
> I read somewhere, it is still 3.10.2, which adds a friction point:
> https://graspingtech.com/upgrade-cmake/
> 
> > The version of CMake installed by APT on Ubuntu 18.04 is currently 3.10.2.

Since refactoring to support version 3.10 would require extra effort and Storage team sync, I would like to first fix the fact that storage is asking for version 3.15 and core for 3.12, since that's not good to delay more time on master.

Later we can define if we want to make more changes to support old versions.

**Review by antkmsft — APPROVED** [review]

**Comment by ahsonkhan** [comment]
> Later we can define if we want to make more changes to support old versions.

Please file an issue.

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,7 +1,7 @@
 # Copyright (c) Microsoft Corporation. All rights reserved.
 # SPDX-License-Identifier: MIT
 
-cmake_minimum_required (VERSION 3.15)
+cmake_minimum_required (VERSION 3.13)
```
Comment (ahsonkhan) [inline CMakeLists.txt:4]: Looks like we are going with 3.13 across the board. Update PR title?

**Pull request merged**